src/Microsoft.Android.Sdk.ILLink/TypeMappingStep.cs

```diff
@@ -77,22 +77,47 @@ protected override void EndProcess ()
 			 */
 			il.Emit (Mono.Cecil.Cil.OpCodes.Ldarg_0);
 			il.Emit (Mono.Cecil.Cil.OpCodes.Ldfld, field);
-			il.Emit (Mono.Cecil.Cil.OpCodes.Ldstr, javaKey);
-			il.Emit (Mono.Cecil.Cil.OpCodes.Ldtoken, module.ImportReference (typeDefinition));
+			il.Emit (Mono.Cecil.Cil.OpCodes.Ldstr, javaName);
+			il.Emit (Mono.Cecil.Cil.OpCodes.Ldtoken, module.ImportReference (SelectTypeDefinition (javaName, list)));
 			il.Emit (Mono.Cecil.Cil.OpCodes.Call, getTypeFromHandle);
 			il.Emit (Mono.Cecil.Cil.OpCodes.Callvirt, addMethod);
 		}
 
 		il.Emit (Mono.Cecil.Cil.OpCodes.Ret);
 	}
 
+	TypeDefinition SelectTypeDefinition (string javaName, List<TypeDefinition> list)
+	{
+		if (list.Count == 1)
+			return list[0];
+
+		var best = list[0];
+		foreach (var type in list) {
+			if (type == best)
+				continue;
+			if ((type.IsAbstract || type.IsInterface) &&
+					!best.IsAbstract &&
+					!best.IsInterface &&
+					type.IsAssignableFrom (best, Context)) {
+				best = type;
+			}
+		}
```

Comment:
This code is trying to be the same as:

https://github.com/dotnet/android/blob/0fb7b453e913dcbe24d9784c37d2d68513e59633/src/Xamarin.Android.Build.Tasks/Utilities/TypeMapGenerator.cs#L314-L321